src/number_theory/modular.lean

```diff
@@ -372,6 +379,343 @@ begin
       simp [T', sub_eq_add_neg] } }
 end
 
+/-- The standard open fundamental domain of the action of `SL(2,ℤ)` on `ℍ` -/
+def fundamental_domain_open : set ℍ :=
+{z | 1 < ((z:ℂ).norm_sq) ∧ |z.re| < (1 : ℝ) / 2}
+
+notation `𝒟ᵒ` := fundamental_domain_open
+
+/-- MOVE TO INT SOMEWHERE -/
+lemma int.eq_one_or_neg_one_of_mul_eq_one {z w : ℤ} (h : z * w = 1) : z = 1 ∨ z = -1 :=
+int.is_unit_iff.mp (is_unit_of_mul_eq_one z w h)
+
+lemma int.eq_one_or_neg_one_of_mul_eq_one' {z w : ℤ} (h : z * w = 1) : (z = 1 ∧ w = 1) ∨
+(z = -1 ∧ w = -1) :=
+begin
+  cases int.eq_one_or_neg_one_of_mul_eq_one h,
```

Comment:
Since `units ℤ` is a fintype, I think you can prove this by `dec_trivial` after lifting `z` and `w` to `units ℤ`. That should give a 3 line proof.

Reply:
So we tried this, but the `can_lift` instance for units requires a `group_with_zero`, which `ℤ` isn't. Hopefully in the future, someone will add this functionality and then this proof can be simplified...?